client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/IndicesPrivileges.java

```diff
@@ -0,0 +1,287 @@
+                        .collect(Collectors.toList());
+                final Tuple<Collection<String>, Collection<String>> fields =
+                        (Tuple<Collection<String>, Collection<String>>) constructorObjects[i++];
+                final String query = (String) constructorObjects[i];
+                return new IndicesPrivileges(indices, privileges, fields.v1(), fields.v2(), query);
+            });
+
+    static {
+        @SuppressWarnings("unchecked")
+        final ConstructingObjectParser<Tuple<Collection<String>, Collection<String>>, Void> fls_parser =
+                new ConstructingObjectParser<>( "field_level_parser", false, constructorObjects -> {
+                        int i = 0;
+                        final Collection<String> grantFields = (Collection<String>) constructorObjects[i++];
+                        final Collection<String> exceptFields = (Collection<String>) constructorObjects[i];
+                        return new Tuple<>(grantFields, exceptFields);
+                    });
+        fls_parser.declareStringArray(optionalConstructorArg(), GRANT_FIELDS);
+        fls_parser.declareStringArray(optionalConstructorArg(), EXCEPT_FIELDS);
+
+        PARSER.declareStringArray(constructorArg(), NAMES);
+        PARSER.declareStringArray(constructorArg(), PRIVILEGES);
+        PARSER.declareObject(optionalConstructorArg(), fls_parser, FIELD_PERMISSIONS);
+        PARSER.declareStringOrNull(optionalConstructorArg(), QUERY);
+    }
+
+    private final Set<String> indices;
+    private final Set<IndexPrivilege> privileges;
+    // '*' means all fields (default value), empty means no fields
+    private final Set<String> grantedFields;
+    // empty means no field is denied
+    private final Set<String> deniedFields;
+    // missing query means all documents, i.e. no restrictions
+    private final @Nullable String query;
+
+    private IndicesPrivileges(Collection<String> indices, Collection<IndexPrivilege> privileges, @Nullable Collection<String> grantedFields,
+            @Nullable Collection<String> deniedFields, @Nullable String query) {
+        // we do all null checks inside the constructor
+        if (null == indices || indices.isEmpty()) {
+            throw new IllegalArgumentException("indices privileges must refer to at least one index name or index name pattern");
+        }
+        if (null == privileges || privileges.isEmpty()) {
+            throw new IllegalArgumentException("indices privileges must define at least one privilege");
+        }
+        this.indices = Collections.unmodifiableSet(new HashSet<>(indices));
+        this.privileges = Collections.unmodifiableSet(new HashSet<>(privileges));
+        // all fields granted unless otherwise specified
+        this.grantedFields = Collections
+                .unmodifiableSet(grantedFields != null ? new HashSet<>(grantedFields) : Collections.singleton("*"));
```

Comment:
`['*']` and `null` are equivalent on the server. That is, two `privileges` otherwise equal but with only this field different in this way, are equal on the server. And I think they should be equal on the client side as well. If we agree to this, and we don't want to touch the response, then this equivalence has to be accounted for in the `equals` method.

Reply:
@hub-cap there is no conversion on the transport action either. Just that the *null* and \* are equivalent values for the `granted_fields`.

I have refrained from doing this conversion. In this case there might be `IndicesPrivileges` that report `false` for *equals* but who are indeed equivalent.

Testing equivalence of `IndicesPrivileges` is a higher client's business.